src/main/java/com/commercetools/sync/commons/helpers/CustomReferenceResolver.java

```diff
@@ -48,11 +52,47 @@ protected CustomReferenceResolver(@Nonnull final S options, @Nonnull final TypeS
      *
      * @param draftBuilder the draft builder to resolve it's references.
      * @return a {@link CompletionStage} that contains as a result a new draft instance with resolved custom
-     *      type references or, in case an error occurs during reference resolution,
-     *      a {@link ReferenceResolutionException}.
+     *         type references or, in case an error occurs during reference resolution,
+     *         a {@link ReferenceResolutionException}.
      */
     protected abstract CompletionStage<B> resolveCustomTypeReference(@Nonnull B draftBuilder);
 
+    /**
+     * Given a draft of {@code D} (e.g. {@link CategoryDraft}) this method attempts to resolve it's custom type
+     * reference to return {@link CompletionStage} which contains a new instance of the draft with the resolved
+     * custom type reference. The key of the custom type is taken from the from the id field of the reference.
+     *
+     * <p>The method then tries to fetch the key of the custom type, optimistically from a
+     * cache. If the key is is not found, the resultant draft would remain exactly the same as the passed
+     * draft (without a custom type reference resolution).
+     *
+     * @param draftBuilder the draft builder to resolve it's references.
+     * @param customGetter a function to return the CustomFieldsDraft instance of the draft builder.
+     * @param customSetter a function to set the CustomFieldsDraft instance of the builder and return this builder.
+     * @param errorMessage the error message to inject in the {@link ReferenceResolutionException} if it occurs.
+     * @return a {@link CompletionStage} that contains as a result a new draft instance with resolved custom
+     *         type references or, in case an error occurs during reference resolution,
+     *         a {@link ReferenceResolutionException}.
+     */
+    @Nonnull
+    protected CompletionStage<B> resolveCustomTypeReference(
+        @Nonnull final B draftBuilder,
+        @Nonnull final Function<B, CustomFieldsDraft> customGetter,
+        @Nonnull final BiFunction<B, CustomFieldsDraft, B> customSetter,
+        @Nonnull final String errorMessage) {
+
+        final CustomFieldsDraft custom = customGetter.apply(draftBuilder);
```

Comment:
actually it's "custom" on purpose because that's how the field is named from on our API :) : 

examples: 
https://docs.commercetools.com/http-api-types.html#assetdraft
https://docs.commercetools.com/http-api-projects-channels.html#channeldraft